components/productList/FilteredProductList/FilteredProductList.tsx

```diff
@@ -0,0 +1,216 @@
+import { queryTypes, TransitionOptions, useQueryState } from "next-usequerystate";
+import { useEffect, useState } from "react";
+
+import { ProductCollection } from "@/components/ProductCollection";
+import { translate } from "@/lib/translations";
+import {
+  AttributeFilterFragment,
+  OrderDirection,
+  ProductFilterInput,
+  ProductOrderField,
+} from "@/saleor/api";
+
+import {
+  getPillsData,
+  parseQueryAttributeFilters,
+  serializeQueryAttributeFilters,
+} from "./attributes";
+import { FilterDropdown } from "./FilterDropdown";
+import { FilterPill, FilterPills } from "./FilterPills";
+import { parseQuerySort, serializeQuerySort, UrlSorting } from "./sorting";
+import { SortingDropdown } from "./SortingDropdown";
+import { StockToggle } from "./StockToggle";
+
+export interface FilteredProductListProps {
+  attributeFiltersData: AttributeFilterFragment[];
+  collectionIDs?: string[];
+  categoryIDs?: string[];
+}
+
+export interface Filters {
+  sortBy: string;
+  attributes: Record<string, Array<string>>;
+}
+
+export interface ListFilters {
+  attributes: {};
+}
+
+export function FilteredProductList({
+  attributeFiltersData,
+  collectionIDs,
+  categoryIDs,
+}: FilteredProductListProps) {
+  const [queryFilters, setQueryFilters] = useQueryState("filters", {
+    parse: parseQueryAttributeFilters,
+    serialize: serializeQueryAttributeFilters,
+    defaultValue: [],
+  });
+
+  const [itemsCounter, setItemsCounter] = useState(0);
+
+  const [sortByQuery, setSortByQuery] = useQueryState("sortBy", {});
+
+  const sortBy = parseQuerySort(sortByQuery);
+  const setSortBy = (
+    value: UrlSorting | undefined | null,
+    transitionOptions?: TransitionOptions | undefined
+  ) => setSortByQuery(serializeQuerySort(value), transitionOptions);
+
+  const [inStockFilter, setInStockFilter] = useQueryState(
+    "inStock",
+    queryTypes.boolean.withDefault(false)
+  );
+
+  const [productsFilter, setProductsFilter] = useState<ProductFilterInput>();
+
+  const pills: FilterPill[] = getPillsData(queryFilters, attributeFiltersData);
+
+  useEffect(() => {
+    setProductsFilter({
+      attributes: queryFilters.filter((f) => f.values?.length),
+      ...(categoryIDs?.length && { categories: categoryIDs }),
+      ...(collectionIDs?.length && { collections: collectionIDs }),
+      ...(inStockFilter && { stockAvailability: "IN_STOCK" }),
+    });
+    // Eslint does not recognize stringified queryFilters, so we have to ignore it
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [inStockFilter, JSON.stringify(queryFilters), categoryIDs, collectionIDs]);
+
+  const removeAttributeFilter = (attributeSlug: string, choiceSlug: string) => {
```

Comment:
All these helper functions could potentially either be in utils or moved to a hook (whatever is necessary) so this component doesn't keep all this mapping and filtering logic. Just a suggestion tho